torch/distributed/distributed_c10d.py

```diff
@@ -2797,3 +2791,124 @@ def new_group(ranks=None, timeout=default_pg_timeout, backend=None, pg_options=N
+            will provide errors to the user which can be caught and handled,
+            but due to its blocking nature, it has a performance overhead. On
+            the other hand, ``NCCL_ASYNC_ERROR_HANDLING`` has very little
+            performance overhead, but crashes the process on errors. This is
+            done since CUDA execution is async and it is no longer safe to
+            continue executing user code since failed async NCCL operations
+            might result in subsequent CUDA operations running on corrupted
+            data. Only one of these two environment variables should be set.
+        backend (str or Backend, optional): The backend to use. Depending on
+            build-time configurations, valid values are ``gloo`` and ``nccl``.
+            By default uses the same backend as the global group. This field
+            should be given as a lowercase string (e.g., ``"gloo"``), which can
+            also be accessed via :class:`Backend` attributes (e.g.,
+            ``Backend.GLOO``). If ``None`` is passed in, the backend
+            corresponding to the default process group will be used. Default is
+            ``None``.
+        pg_options (ProcessGroupOptions, optional): process group options
+            specifying what additional options need to be passed in during
+            the construction of specific process groups. i.e. for the ``nccl``
+            backend, is_high_priority_stream can be specified so that process
+            group can pick up high priority cuda streams.
+
+    Returns:
+        The subgroup containing the current rank, and all the subgroups used for cleanup.
+    """
+    subgroups = []
+    cur_subgroup = None
+    if ranks_per_subgroup_list:
+        for ranks in ranks_per_subgroup_list:
+            subgroup = new_group(
+                ranks=ranks, timeout=timeout, backend=backend, pg_options=pg_options
+            )
+            subgroups.append(subgroup)
+
+            rank = get_rank()
+            if rank in ranks:
+                cur_subgroup = subgroup
+                logger.info("Rank {} is assigned to subgroup {}".format(rank, ranks))
+        return cur_subgroup, subgroups
+
+    # By default, create one subgroup per machine.
+    # Assume that each machine has the same number of devices.
+    num_devices_per_machine = (
+        torch.cuda.device_count()
+        if torch.cuda.is_available()
+        else (os.cpu_count() or 1)
+    )
+    # If the given main group does not use all the devices,
+    # then the number of devices used by the subgroup should exceed the world size.
+    num_devices_per_machine = min(get_world_size(group), num_devices_per_machine)
+    for machine_rank in range(get_world_size(group) // num_devices_per_machine):
+        start_rank = machine_rank * num_devices_per_machine
+        end_rank = start_rank + num_devices_per_machine
+        per_machine_ranks = list(range(start_rank, end_rank))
+        subgroup = new_group(
```

Comment:
if `ranks_per_subgroup_list=None`, then we can do the above computation to compute per machine ranks list, and then re-use the above logic that calls `new_group` and eliminate some of the code duplication here.

Reply:
If `ranks_per_subgroup_list=None`, then the first if statement will not be executed, and per-machine subgroups will be created.

I don't see much code duplication in the implementations for these 2 cases. Which lines of code do you want to deduplicate?